File: bin/MadGraph5_aMCatNLO/gridpack_test/BulkGraviton_hh_GF_HH_narrow_M900_gridpack/config/SCRAM/find-deps-tree.pl
#!/usr/bin/env perl

BEGIN{unshift @INC,$ENV{SCRAM_TOOL_HOME};}

use Cache::CacheUtilities;

my @conf=();

my $pack=shift;
my $dep=shift;

my $base=shift;  #CMSSW_BASE
push @conf,$base;

$base=shift || "";  #CMSSW_RELEASE_BASE
if ($base ne ""){push @conf,$base;}

$base=shift || "";  #FULL_RELEASE_BASE for patch releases
if ($base ne ""){push @conf,$base;}

my %cache=();
foreach my $p (@conf){&readDeps($p, \%cache);}
my $toolfile=$conf[0]."/.SCRAM/$ENV{SCRAM_ARCH}/ToolCache.db.gz";
my $cc=&Cache::CacheUtilities::read($toolfile);
foreach my $tool (keys %{$cc->{SETUP}})
{
  $cache{$tool}={};
  foreach my $d (@{$cc->{SETUP}{$tool}{USE}}){$cache{$tool}{$d}=1;}
}

foreach my $p (keys %cache)
{
  if ($p=~/^(.+)\/src$/)
  {
    my $pp=$1;
    if (!exists $cache{$pp}){$cache{$pp}={};}
    foreach my $d (keys %{$cache{$p}}){$cache{$pp}{$d}=1;}
    delete $cache{$p};
  }
}

if (!exists $cache{$pack}){print "Error: No such package/tool: $pack\n";}
if (!exists $cache{$dep}){print "Error: No such package/tool: $dep\n";}
my @chain=();
&findDeps(\%cache,$pack,$dep, \@chain);


sub readDeps()
{
  my ($release, $cache)=@_;
  my $cfile="${release}/.SCRAM/$ENV{SCRAM_ARCH}/ProjectCache.db.gz";
  if (!-f $cfile){return;}
  my $cc=&Cache::CacheUtilities::read($cfile);
  foreach my $dir (keys %{$cc->{BUILDTREE}})
  {
    if (-e $cache->{$dir}){next;}
    $cache->{$dir}={};
    foreach my $d (keys %{$cc->{BUILDTREE}{$dir}{RAWDATA}{DEPENDENCIES}}){$cache->{$dir}{$d}=1;}
  }
}

sub findDeps()
{
  my ($cache, $pack, $dep, $chain)=@_;
  push @$chain,$pack;
  foreach my $d (sort keys %{$cache{$pack}})
  {
    if ($d eq $dep){print "  ",join("->",@{$chain}),"->$d\n";}
    else
    {
      &findDeps($cache,$d,$dep, $chain);
      if (exists $cache{"${d}/src"}){&findDeps($cache,"${d}/src",$dep, $chain);}
    }
  }
  pop @$chain;
}

File: bin/MadGraph5_aMCatNLO/ewkwpjjjj_5f_LO/ewkwpjjjj_5f_LO_gridpack/config/SCRAM/find-deps-tree.py
#!/usr/bin/env python3
import sys
from os import environ
from os.path import join, exists, isdir
from subprocess import getstatusoutput
sys.path.append(environ["SCRAM_TOOL_HOME"])

from SCRAM.Core.Utils import readProducts
from SCRAM.BuildSystem.ToolManager import ToolManager
from SCRAM.Configuration.ConfigArea import ConfigArea

def addDeps(cache, product):
    if "USE" in product:
        for use in product["USE"]:
            cache[use] = 1
    return

def readDeps(release, cache):
    area = ConfigArea()
    area.location(release)
    pcache = readProducts(area)
    for dir in pcache:
        if dir in cache:
            continue
        cache[dir] = {}
        addDeps(cache[dir], pcache[dir])
        if "BUILDPRODUCTS" in pcache[dir]:
            for type in pcache[dir]["BUILDPRODUCTS"]:
                for prod in pcache[dir]["BUILDPRODUCTS"][type]:
                    addDeps(cache[dir], pcache[dir]["BUILDPRODUCTS"][type][prod])
    return

def findDeps(cache, pack, dep, chain):
    chain.append(pack)
    if pack in cache:
        for use in sorted(cache[pack].keys()):
            if use == dep:
                print ("  %s->%s" % ("->".join(chain), use))
            else:
                findDeps(cache, use, dep, chain)
                use = "%s/src" % use
                if use in cache:
                    findDeps(cache, use, dep, chain)
    del chain[-1]
    return

pack=sys.argv[1]
dep=sys.argv[2]
cache = {}

area = ConfigArea()
area.location(sys.argv[3])
toolmgr = ToolManager(area)
tools = toolmgr.loadtools()

for base in sys.argv[3:]:
    if base and isdir(base):
        readDeps(base, cache)

for tool in tools:
    cache[tool] = {}
    if "USE" in tools[tool]:
        for use in tools[tool]["USE"]:
            cache[tool][use] = 1

for dir in list(cache.keys()):
    if dir.endswith("/src"):
        pdir = dir[:-4]
        if pdir not in cache:
            cache[pdir] = {}
        for use in cache[dir]:
            cache[pdir][use] = 1
        del cache[dir]

if pack not in cache:
    print ("Error: No such package/tool: %s" % pack)
if dep not in cache:
    print ("Error: No such package/tool: %s" % dep)
findDeps(cache, pack, dep, [])
